File: FinalProject/2_GraphAndTree/Baitoan5/Prob15/O1Representation.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node* parent;
    Node* firstChild;
    Node* nextSibling;
    int child_count;
    vector<Node*> children_list;

    Node(int val) {
        data=val;
        parent=nullptr;
        firstChild=nullptr;
        nextSibling=nullptr;
        child_count=0;
    }
};

class Tree {
private:
    Node* root;

public:
    Tree(int root_val) {
        root=new Node(root_val);
    }

    Node* get_root() {
        return root;
    }

    Node* add_child(Node* parent, int val) {
        Node* child=new Node(val);
        child->parent=parent;

        //gắn vào danh sách con
        if (!parent->firstChild)
            parent->firstChild=child;
        else {
            Node* temp=parent->firstChild;
            while (temp->nextSibling)
                temp=temp->nextSibling;
            temp->nextSibling=child;
        }

        //cập nhật child count và children list
        parent->child_count++;
        parent->children_list.push_back(child);

        return child;
    }

    int number_of_children(Node* node) {
        return node->child_count;
    }

    vector<Node*> children(Node* node) {
        return node->children_list;
    }

    void print_children(Node* node) {
        cout<<"Children of "<<node->data<<": ";
        for (Node* child : node->children_list)
            cout<<child->data<<" ";
        cout<<"\n";
    }
};

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    Tree T(1); //root là 1
    Node* root=T.get_root();

    Node* a=T.add_child(root, 2);
    Node* b=T.add_child(root, 3);
    Node* c=T.add_child(root, 4);

    T.add_child(a, 5);
    T.add_child(a, 6);

    cout<<"Number of children of root: "<<T.number_of_children(root)<<"\n";
    cout<<"Number of children of node 2: "<<T.number_of_children(a)<<"\n";

    T.print_children(root);
    T.print_children(a);

    return 0;
}

File: FinalProject/2_GraphAndTree/Baitoan5/Ex110/O1Representation.py
class ExtendedFCNS:
    def __init__(self, n):
        self.n = n
        self.root = None  #lưu chỉ số root
        self.first_child = [None] * n  #first_child[v] = con đầu tiên của v
        self.next_sibling = [None] * n  #next_sibling[v] = sibling kế tiếp của v
        self.num_children = [0] * n  #số con của mỗi node
        self.parent = [None] * n

    def set_root(self, r):
        self.root = r

    def add_child(self, parent, child):
        #thêm child vào đầu danh sách con của parent
        self.next_sibling[child] = self.first_child[parent]
        self.first_child[parent] = child
        self.parent[child] = parent
        self.num_children[parent] += 1

    def root_node(self):
        return self.root

    def number_of_children(self, v):
        return self.num_children[v]

    def children(self, v):
        #trả về con trỏ tới con đầu tiên của v
        return self.first_child[v]

    #duyệt các con của node v
    def iterate_children(self, v):
        c = self.first_child[v]
        while c is not None:
            yield c
            c = self.next_sibling[c]

#ví dụ
if __name__ == "__main__":
    tree = ExtendedFCNS(7)
    tree.set_root(0)
    tree.add_child(0, 1)
    tree.add_child(0, 2)
    tree.add_child(1, 3)
    tree.add_child(1, 4)
    tree.add_child(2, 5)
    tree.add_child(2, 6)

    print("Root:", tree.root_node())
    print("Number of children of node 0:", tree.number_of_children(0))
    print("Children of node 0:", list(tree.iterate_children(0)))
    print("Number of children of node 1:", tree.number_of_children(1))
    print("Children of node 1:", list(tree.iterate_children(1)))
